### PR TITLE
Extend HumanEntity#dropItem API

### LINKED ISSUE #11656: Player#dropItem(ItemStack)
### Is your feature request related to a problem?

Currently, there is only a `Player#dropItem(boolean)` method to drop the held item of the player, so when dropping an item from the inventory from a player you have to implement the entire drop logic yourself _or use nms_

### Describe the solution you'd like.

A `Player#dropItem(ItemStack)` method
and maybe even a `Player#dropItem(EquipmentSlot)` method

### Describe alternatives you've considered.

Using NMS or implementing dropping yourself

### Other

My best attempt at implementing it myself (basically what vanilla does)
```java
player.getInventory().remove(item);
var location = player.getEyeLocation().clone();
location.setY(location.getY() - 0.30000001192092896D);
player.getWorld().dropItem(location, item, drop -> {

    var random = ThreadLocalRandom.current();

    var f1 = Mth.sin(player.getPitch() * 0.017453292F);
    var f2 = Mth.cos(player.getPitch() * 0.017453292F);
    var f3 = Mth.sin(player.getYaw() * 0.017453292F);
    var f4 = Mth.cos(player.getYaw() * 0.017453292F);
    var f5 = random.nextFloat() * 6.2831855F;
    var f6 = 0.02F * random.nextFloat();

    var x = (double) (f4 * f2 * 0.3F) + Math.sin(f5) * (double) f6;
    var y = (double) -f1 * 0.3F + 0.1F + (random.nextFloat() - random.nextFloat()) * 0.1F;
    var z = (double) (-f3 * f2 * 0.3F) + Math.cos(f5) * (double) f6;
    drop.setVelocity(new Vector(z, y, x));

    drop.setThrower(player.getUniqueId());
    drop.setPickupDelay(40);
});
```
or with NMS
```java
((CraftPlayer) player).getHandle().drop(((CraftItemStack) item).handle, false, true, true);
```
Additionally, to make it seem like the player dropped the item themselves, swing their main hand
```java
player.swingMainHand();
```

### PR DESCRIPTION
**Pull request by Strokkur424**
Closes #11656

# Motive
Developers currently have to either use NMS or copy-paste the already existing NMS implementation. This PR aims to fix that by exposing ways to call that NMS implementation more easily by extending onto the (Human)Player#dropItem method to provide more extensive ways of making the player drop items.

## What does it add?
The following overloads to the `HumanPlayer#dropItem(...)` method have been added:
* `Item dropItem(ItemStack itemStack)`
* `Item dropItem(ItemStack itemStack, boolean persistThrower)`
* `Item dropItem(ItemStack itemStack, boolean persistThrower, boolean throwRandomly)`
* `Item dropItem(int slot)`
* `Item dropItem(int slot, boolean persistThrower)`
* `Item dropItem(int slot, boolean persistThrower, boolean throwRandomly)`
* `Item dropItem(EquipmentSlot slot)`
* `Item dropItem(EquipmentSlot slot, boolean persistThrower)`
* `Item dropItem(EquipmentSlot slot, boolean persistThrower, boolean throwRandomly)`

## Implementation details
Internally, the implementation calls `net.minecraft.world.entity.player.Player#dropItem(...)`, making it a fairly maintainable API. But as that method does not remove the item, I have added that by calling `this.inventory.setItem(slot, null)`.

## JavaDocs status
I have added simple JavaDocs, but the methods should be fairly self explanatory.

# Usage example
```java
/**
 * Makes the player drop an item of the type they using. Does not cancel the event
 */
@EventHandler
public void onInteract(final PlayerInteractEvent event) {
    event.getPlayer().dropItem(event.getItem());
    event.setCancelled(true);
}

/**
 * Strips the player of all currently wearing armor
 */
private void stripOfArmor(final Player player) {
    player.dropItem(EquipmentSlot.HEAD);
    player.dropItem(EquipmentSlot.CHEST);
    player.dropItem(EquipmentSlot.LEGS);
    player.dropItem(EquipmentSlot.FEET);
}
```

> [!NOTE]
> I am not 100% certain that the methods work in every possible scenario, as I don't know how to test that. If somebody finds a bug, make sure to tell me and I will make sure to fix it.

### CODE REVIEW

**Comment by NonSwag** [comment]
Thanks for that PR, I was too lazy to do it myself :smile: 
Internally, the item is already being removed from the player's inventory by setting its type to air and amount to 0 so you don't need to remove it yourself though

**Comment by NonSwag** [comment]
I don't think slot related methods are a good idea
having methods for itemstack and equipment slots cover basically all needs

File: patches/server/1073-Extend-HumanEntity-dropItem-API.patch
Comment (NonSwag) [inline patches/server/1073-Extend-HumanEntity-dropItem-API.patch]: I am not sure if it is a smart idea to swing the main hand internally
maybe in some cases, the user of this method wants the player to swing their offhand so it should be up to them to do such a thing

File: patches/server/1073-Extend-HumanEntity-dropItem-API.patch
Comment (NonSwag) [inline patches/server/1073-Extend-HumanEntity-dropItem-API.patch]: Why not on the server player themselves?

File: patches/server/1073-Extend-HumanEntity-dropItem-API.patch
Comment (NonSwag) [inline patches/server/1073-Extend-HumanEntity-dropItem-API.patch:80]: Probably an exception would be more appropriate

File: patches/server/1073-Extend-HumanEntity-dropItem-API.patch
Comment (NonSwag) [inline patches/server/1073-Extend-HumanEntity-dropItem-API.patch:49]: If this method stays (see my previous comment) an exception would be more appropriate

**Comment by Strokkur424** [comment]
> I don't think slot related methods are a good idea having methods for itemstack and equipment slots cover basically all needs

The issue with that is that dropping by ItemStack pretty much always forces the **first** ItemStack to be dropped. Dropping by slot allows to choose an explicit one, like the one current held, as example

File: patches/api/0501-Extend-HumanEntity-dropItem-API.patch
Comment (NonSwag) [inline patches/api/0501-Extend-HumanEntity-dropItem-API.patch]: retainOwnership is very misleading (from nms itself) it is not setting the owner but the thrower
the difference here is that if an item is owned it can only be picked up by this very owner
the thrower is just the entity's uuid that threw the item
potentially a less misleading name would be appropriate or clarification in the docs

File: patches/api/0501-Extend-HumanEntity-dropItem-API.patch
Comment (NonSwag) [inline patches/api/0501-Extend-HumanEntity-dropItem-API.patch]: same here

File: patches/api/0501-Extend-HumanEntity-dropItem-API.patch
Comment (NonSwag) [inline patches/api/0501-Extend-HumanEntity-dropItem-API.patch]: same here

File: patches/api/0501-Extend-HumanEntity-dropItem-API.patch
Comment (NonSwag) [inline patches/api/0501-Extend-HumanEntity-dropItem-API.patch]: here

File: patches/api/0501-Extend-HumanEntity-dropItem-API.patch
Comment (NonSwag) [inline patches/api/0501-Extend-HumanEntity-dropItem-API.patch]: here

**Comment by Strokkur424** [comment]
> Internally, the item is already being removed from the player's inventory by setting its type to air and amount to 0 so you don't need to remove it yourself though

That sadly not the case for items equipped on armor. I am not quite sure why, but in my testing, armor items did not get cleared by themselves

File: patches/api/0501-Extend-HumanEntity-dropItem-API.patch
Comment (NonSwag) [inline patches/api/0501-Extend-HumanEntity-dropItem-API.patch]: here

File: patches/api/0501-Extend-HumanEntity-dropItem-API.patch
Comment (Strokkur424) [reply patches/api/0501-Extend-HumanEntity-dropItem-API.patch]: Ah okay, thanks for clarifying that. It mislead even me 😅

**Comment by NonSwag** [comment]
> > Internally, the item is already being removed from the player's inventory by setting its type to air and amount to 0 so you don't need to remove it yourself though
> 
> That sadly not the case for items equipped on armor. I am not quite sure why, but in my testing, armor items did not get cleared by themselves

in my tests it worked (1.21.1 and 1.21.3)

**Comment by NonSwag** [comment]
> > I don't think slot related methods are a good idea having methods for itemstack and equipment slots cover basically all needs
> 
> The issue with that is that dropping by ItemStack pretty much always forces the **first** ItemStack to be dropped. Dropping by slot allows to choose an explicit one, like the one current held, as example

I did not think of that
in that case, it totally makes sense

File: patches/server/1073-Extend-HumanEntity-dropItem-API.patch
Comment (Strokkur424) [reply patches/server/1073-Extend-HumanEntity-dropItem-API.patch:80]: I would rather make it return null, as in no item was found, so no item was dropped. Because otherwise you'd have to manually run a check for whether the item actually exists, and it just turns into a mess

Comment (NonSwag) [reply patches/server/1073-Extend-HumanEntity-dropItem-API.patch:80]: valid point

File: patches/server/1073-Extend-HumanEntity-dropItem-API.patch
Comment (Strokkur424) [reply patches/server/1073-Extend-HumanEntity-dropItem-API.patch]: Ah, the object returned by `this.getHandle()` already has the method definition for `#drop(...)`. I kinda just looked at the one `#dropItem(boolean)` method and they had used a ServerPlayer as well, so I kind of assumed the nms player to not have that method

**Comment by Strokkur424** [comment]
> in my tests it worked (1.21.1 and 1.21.3)

When I remove that line, it turns into an dupe glitch 😥

https://github.com/user-attachments/assets/86db1db4-a9ba-4c8a-9ac8-1acd7d08951e

```java
@EventHandler
public void onInventoryClick(final InventoryClickEvent event) {
    if (!event.getClickedInventory().equals(event.getWhoClicked().getInventory())) {
        return;
    }

    event.setCancelled(true);
    event.getWhoClicked().dropItem(event.getSlot());
}
```

**Comment by NonSwag** [comment]
hmm i see

**Comment by Strokkur424** [comment]
Oops, the patch has a problem. Give me a second to fix it

File: patches/api/0501-Extend-HumanEntity-dropItem-API.patch
Comment (NonSwag) [inline patches/api/0501-Extend-HumanEntity-dropItem-API.patch]: maybe "persistThrower" would be a more fitting name

Comment (Strokkur424) [reply patches/api/0501-Extend-HumanEntity-dropItem-API.patch]: Changed!

File: patches/server/1073-Extend-HumanEntity-dropItem-API.patch
Comment (Strokkur424) [reply patches/server/1073-Extend-HumanEntity-dropItem-API.patch:49]: Also just to note: As I am using this method when dropping via ItemStack, this would also throw an exception each time the ItemStack is not found, which seems counterproductive. So same case as the other one

Comment (NonSwag) [reply patches/server/1073-Extend-HumanEntity-dropItem-API.patch:49]: In that case of the slot is outside the inventory bounds we should definitely throw

Comment (Strokkur424) [reply patches/server/1073-Extend-HumanEntity-dropItem-API.patch:49]: The issue is:
1. I completely agree with you on user input. If you put `137` as the slot, is should throw an `IndexOutOfBoundsException`
2. But if they put in an ItemStack that hasn't been found, an `IndexOutOfBoundsException` feels weird

Comment (Strokkur424) [reply patches/server/1073-Extend-HumanEntity-dropItem-API.patch:49]: Maybe I can add a check inside the ItemStack overload one to check if the ItemStack exists, and only then pass the slot in?

Comment (Strokkur424) [reply patches/server/1073-Extend-HumanEntity-dropItem-API.patch:49]: Yeah okay so I made it now that if you pass an incorrect slot in, it throws an `IndexOutOfBoundsException`, like this:
```
[20:14:39 ERROR]: Command exception: /drop -1                                                                                                                                                           
java.lang.IndexOutOfBoundsException: Slot -1 out of range for inventory of size 41                                                                                                                      
        at org.bukkit.craftbukkit.entity.CraftHumanEntity.dropItem(CraftHumanEntity.java:839) ~[main/:?]
        at org.bukkit.craftbukkit.entity.CraftHumanEntity.dropItem(CraftHumanEntity.java:832) ~[main/:?]
        at test-plugin-1.0.0-SNAPSHOT.jar/io.papermc.testplugin.TestPlugin.logic(TestPlugin.java:29) ~[test-plugin-1.0.0-SNAPSHOT.jar:?]
        at test-plugin-1.0.0-SNAPSHOT.jar/io.papermc.testplugin.TestPlugin$1.execute(TestPlugin.java:21) ~[test-plugin-1.0.0-SNAPSHOT.jar:?]
        at io.papermc.paper.command.brigadier.bukkit.BukkitCommandNode$BukkitBrigCommand.run(BukkitCommandNode.java:82) ~[main/:?]
        at com.mojang.brigadier.context.ContextChain.runExecutable(ContextChain.java:73) ~[brigadier-1.3.10.jar:?]
        at net.minecraft.commands.execution.tasks.ExecuteCommand.execute(ExecuteCommand.java:31) ~[filteredMinecraft.jar:?]
        at net.minecraft.commands.execution.tasks.ExecuteCommand.execute(ExecuteCommand.java:19) ~[filteredMinecraft.jar:?]
        at net.minecraft.commands.execution.UnboundEntryAction.lambda$bind$0(UnboundEntryAction.java:8) ~[filteredMinecraft.jar:?]
        at net.minecraft.commands.execution.CommandQueueEntry.execute(CommandQueueEntry.java:5) ~[filteredMinecraft.jar:?]
        at net.minecraft.commands.execution.ExecutionContext.runCommandQueue(ExecutionContext.java:103) ~[filteredMinecraft.jar:?]
        at net.minecraft.commands.Commands.executeCommandInContext(Commands.java:448) ~[main/:?]
        at net.minecraft.commands.Commands.performCommand(Commands.java:355) ~[main/:?]
        at net.minecraft.commands.Commands.performCommand(Commands.java:342) ~[main/:?]
        at net.minecraft.commands.Commands.performCommand(Commands.java:337) ~[main/:?]
        at net.minecraft.server.network.ServerGamePacketListenerImpl.performUnsignedChatCommand(ServerGamePacketListenerImpl.java:2241) ~[main/:?]
        at net.minecraft.server.network.ServerGamePacketListenerImpl.lambda$handleChatCommand$11(ServerGamePacketListenerImpl.java:2215) ~[main/:?]
        at net.minecraft.server.TickTask.run(TickTask.java:18) ~[filteredMinecraft.jar:?]
        at net.minecraft.util.thread.BlockableEventLoop.doRunTask(BlockableEventLoop.java:154) ~[main/:?]
        at net.minecraft.util.thread.ReentrantBlockableEventLoop.doRunTask(ReentrantBlockableEventLoop.java:24) ~[filteredMinecraft.jar:?]
        at net.minecraft.server.MinecraftServer.doRunTask(MinecraftServer.java:1577) ~[main/:?]
        at net.minecraft.server.MinecraftServer.doRunTask(MinecraftServer.java:207) ~[main/:?]
        at net.minecraft.util.thread.BlockableEventLoop.pollTask(BlockableEventLoop.java:128) ~[main/:?]
        at net.minecraft.server.MinecraftServer.pollTaskInternal(MinecraftServer.java:1554) ~[main/:?]
        at net.minecraft.server.MinecraftServer.pollTask(MinecraftServer.java:1547) ~[main/:?]
        at net.minecraft.util.thread.BlockableEventLoop.managedBlock(BlockableEventLoop.java:138) ~[main/:?]
        at net.minecraft.server.MinecraftServer.managedBlock(MinecraftServer.java:1498) ~[main/:?]
        at net.minecraft.server.MinecraftServer.waitUntilNextTick(MinecraftServer.java:1508) ~[main/:?]
        at net.minecraft.server.MinecraftServer.runServer(MinecraftServer.java:1344) ~[main/:?]
        at net.minecraft.server.MinecraftServer.lambda$spin$0(MinecraftServer.java:340) ~[main/:?]
        at java.base/java.lang.Thread.run(Thread.java:1583) ~[?:?]
```

**Review by NonSwag — APPROVED** [review]

**Comment by NonSwag** [comment]
Now let's wait for a maintainer for review

**Comment by Malfrador** [comment]
Instead of the `persistThrower` boolean, I would find an optional `UUID` argument allowing you to directly _set_ the thrower more useful. Especially given that method already exists on the resulting `Item` entity anyways. 
Also some Javadoc clarification on what `throwRandomly` does surely would be helpful, without looking at the NMS code I have no clue what that would do.

**Comment by Strokkur424** [comment]
Closing pre softspoon PR in favour of new one

**Pull request closed without merging**
